protocol/kafka_confluent/v2/protocol.go

```diff
@@ -0,0 +1,225 @@
+	}
+	if err := p.applyOptions(opts...); err != nil {
+		return nil, err
+	}
+
+	if p.kafkaConfigMap != nil {
+		if p.consumerTopics != nil && p.consumer == nil {
+			consumer, err := kafka.NewConsumer(p.kafkaConfigMap)
+			if err != nil {
+				return nil, err
+			}
+			p.consumer = consumer
+		}
+		if p.producerDefaultTopic != "" && p.producer == nil {
+			producer, err := kafka.NewProducer(p.kafkaConfigMap)
+			if err != nil {
+				return nil, err
+			}
+			p.producer = producer
+		}
+		if p.producer == nil && p.consumer == nil {
+			return nil, fmt.Errorf("at least receiver or sender topic must be set")
+		}
+	}
+
+	if p.kafkaConfigMap == nil && p.producer == nil && p.consumer == nil {
+		return nil, fmt.Errorf("at least one of the following to initialize the protocol must be set: config, producer, or consumer")
+	}
+	if p.producer != nil {
+		p.producerDeliveryChan = make(chan kafka.Event)
+	}
+	return p, nil
+}
+
+func (p *Protocol) applyOptions(opts ...Option) error {
+	for _, fn := range opts {
+		if err := fn(p); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *Protocol) Send(ctx context.Context, in binding.Message, transformers ...binding.Transformer) (err error) {
+	if p.producer == nil {
+		return fmt.Errorf("producer client must be set")
+	}
+	defer in.Finish(err)
+
+	kafkaMsg := &kafka.Message{
+		TopicPartition: kafka.TopicPartition{
+			Topic:     &p.producerDefaultTopic,
+			Partition: kafka.PartitionAny,
+		},
+	}
+
+	if topic := cecontext.TopicFrom(ctx); topic != "" {
+		kafkaMsg.TopicPartition.Topic = &topic
+	}
+
+	if messageKey := MessageKeyFrom(ctx); messageKey != "" {
+		kafkaMsg.Key = []byte(messageKey)
+	}
+
+	err = WriteProducerMessage(ctx, in, kafkaMsg, transformers...)
+	if err != nil {
+		return err
+	}
+
+	err = p.producer.Produce(kafkaMsg, p.producerDeliveryChan)
+	if err != nil {
+		return err
+	}
+	e := <-p.producerDeliveryChan
+	m := e.(*kafka.Message)
+	if m.TopicPartition.Error != nil {
+		return m.TopicPartition.Error
+	}
+	return nil
+}
+
+func (p *Protocol) OpenInbound(ctx context.Context) error {
+	if p.consumer == nil {
+		return fmt.Errorf("the consumer client must be set")
+	}
+	if p.consumerTopics == nil {
+		return fmt.Errorf("the consumer topics must be set")
+	}
+
+	p.consumerMux.Lock()
+	defer p.consumerMux.Unlock()
+	logger := cecontext.LoggerFrom(ctx)
+
+	// Query committed offsets for each partition
+	if positions := TopicPartitionOffsetsFrom(ctx); positions != nil {
+		if err := p.consumer.Assign(positions); err != nil {
+			return err
+		}
+	}
+
+	logger.Infof("Subscribing to topics: %v", p.consumerTopics)
+	err := p.consumer.SubscribeTopics(p.consumerTopics, p.consumerRebalanceCb)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if !p.consumer.IsClosed() {
+			logger.Infof("Closing consumer %v", p.consumerTopics)
+			if err = p.consumer.Close(); err != nil {
+				logger.Errorf("failed to close the consumer: %v", err)
+			}
+		}
+	}()
+
+	for !p.consumer.IsClosed() {
```

Comment:
simply drop this bool and make it a simple `for { select { } }`. We rely on `ctx` to terminate this loop. Just make sure to protext access to `isClosed` with mutex.

Reply:
Sure! Only keeping the `ctx` to terminate the loop is going to make the logic simpler. However, there is a scenario I'm not sure whether we need to consider:

1. Use the `kafka.Consumer` to initialize a `cloudevents.Client`.  
2. Start polling the message by invoking the `StartReceiver`(blocking) of the `cloudevents.Client`. 
3. If we call the `kafka.Consumer.Close()` outside, then the code will panic causing we can't use a closed consumer to poll message.

Since the `cloudevents.Client` hasn't exposed the `Close()` method, maybe we don't need to handle it? I'm not sure, need your confirm...

Reply:
And I have removed the `p.consumer.IsClosed()` from the consumer for loop now